examples/knative/python/kn-py-vro/handler.py

```diff
@@ -0,0 +1,270 @@
+            if 'Key' in v and 'Value' in v:
+                # pass the key and value to this function recursively to get a correctly structured value
+                vroParam = getVroInputParam((v["Key"], v["Value"]))
+                properties.append({"key": v["Key"], "value": vroParam["value"]})
+        param['value'] = {"properties": {"property": properties}}
+
+    elif value is None:
+        param = None
+
+    else:  # Unhandled data type - try and wrangle to a string
+        app.logger.debug(f'{bgc.WARNING}Unhandled data type "{type(value).__name__}" - forcing to string{bgc.ENDC}')
+        param['type'] = "string"
+        param['value'] = {"string": {"value": str(value)}}
+
+    if type(param) == dict:  # set the parameter name for valid types
+        param["name"] = name[0].lower() + name[1:]
+    return param
+
+
+def get_response(message, code):
+        app.logger.error(message)
+        msg = {
+            'status': code,
+            'error': message,
+        }
+        resp = jsonify(msg)
+        resp.status_code = code
+        return resp
+
+
+app = Flask(__name__)
+@app.route('/', methods=['POST'])
+def handle():
+    """
+    Handles a request to the function
+
+    """
+    global vchost
+
+    # Load the Events that function gets from vCenter through the Event Router
+    app.logger.debug(f'{bgc.HEADER}Reading Cloud Event: {bgc.ENDC}')
+    cevent = from_http(request.headers, request.get_data(),None)
+
+    if cevent._attributes.get("datacontenttype").lower() != "application/json":
+            return get_response(f'invalid datacontenttype for cloud event: {cevent._attributes.get("datacontenttype")}', 500)
+    app.logger.debug(f'{bgc.OKBLUE}Event > {bgc.ENDC}{type(cevent)} - {cevent}')
+
+    try:
+        # CloudEvent - simple validation
+        source = cevent._attributes['source']
+        event = cevent.data
+        _ = event['Key']
+    except KeyError as err:
+        traceback.print_exc(limit=1, file=sys.stderr)  # providing traceback since it helps debug the exact key that failed
+        return get_response(f'Invalid JSON, required key not found > KeyError: {err} > {cevent}', 500)
+    except TypeError as err:
+        traceback.print_exc(limit=1, file=sys.stderr)  # providing traceback since it helps debug the exact key that failed
+        return get_response(f'Invalid JSON, data not iterable > TypeError: {err}', 500)
+
+    app.logger.debug(f'{bgc.HEADER}Validating Input data: {bgc.ENDC}')
+    app.logger.debug(f'{bgc.OKBLUE}Event > {bgc.ENDC}{json.dumps(event, indent=4, sort_keys=True)}')
+
+    vchost = urllib3.util.url.parse_url(source).host
+    app.logger.debug(f'{bgc.HEADER}All tests passed! Build vRO request:{bgc.ENDC}')
+
+    # Build the vRO input parameter object using the event parameters
+    body = {"parameters": []}
+    for item in event.items():
+        # Get the vRO input parameter correctly formatted for the received event data
+        res = getVroInputParam(item)
```

Comment:
question: IMHO `getVroInputParam` can throw an exception in certain cases (e.g. dict lookup). Should we add try/catch around these calls?

Reply:
Agree that there's at least a couple of places in `getVroInputParam` where I've assumed the format of the passed data. If you agree I feel it would be better to add the exception handling in `getVroInputParam` itself.

Reply:
After reviewing the code in the `getVroInputParam` function I decided that I don't think exception handling is required as the couple of places I could see that could cause an exception could be coded a little more defensively.

Let me know what you think of it as it stands to ensure I've not missed anything. My aim here is that any cloud event parameter makes it in to the vRO request somehow, even if it just ends up as a string of JSON as at least then the vRO workflow could handle this.

Reply:
I'm not quite sure I follow why the conversion from CE.date to event and corresponding CE attributes is needed. Same for `event.items()`. I'll wait until you get some unit tests in to see how the code behaves under various conditions.

Reply:
My aim here is to parse each CE.data item and each CE attribute in to vRO input parameters - one for each so they can be picked up in vRO. At the very least I'll change the naming of the `event` variable here as that's not descriptive.